### ISSUE TITLE
Support for Racket 8

### ISSUE THREAD

**Issue by erezgeva** [swig/swig#2830]
Debian racket package in bookworm and up, do not provide any mzdyn object.
Not sure if that is a bug in Debian or racket provides a better solution?

[bookworm racket version 8.7+dfsg1-1](https://packages.debian.org/bookworm/racket), [Files]( https://packages.debian.org/bookworm/amd64/racket/filelist)
[trixie racket version 8.10+dfsg1-2](https://packages.debian.org/trixie/racket), [Files]( https://packages.debian.org/trixie/amd64/racket/filelist)

Old versions have:
/usr/lib/racket/mzdyn3m.o
[bookworm racket version 7.9+dfsg1-2](https://packages.debian.org/bullseye/racket), [Files]( https://packages.debian.org/bullseye/amd64/racket/filelist)

Ubuntu follows. Newer version remove the mzdyn object.
[8.10+dfsg1-2  universe](https://packages.ubuntu.com/noble/racket), [Files](https://packages.ubuntu.com/noble/amd64/racket/filelist)

**Comment by ojwb** [comment]
Racket 8 changed the underlying scheme implementation so SWIG doesn't currently work with this version or newer - see discussion in #920.  We've not had a SWIG/Racket maintainer for a long time but maybe somebody cares and will step forward and fix this.  The plan is to drop Racket support if that doesn't happen and we reach the point where we can no longer test Racket in CI.

**Comment by ojwb** [comment]
This is already documented in the manual BTW:

> SWIG works with Racket versions &lt; 8 (Racket 8 switched to be based on a different scheme interpreter and SWIG hasn't been updated for this change).

```
$ rmadison -uubuntu racket
 racket | 5.3.6+dfsg1-1      | trusty/universe | source, amd64, armhf, i386, powerpc
 racket | 6.3-1              | xenial/universe | source, amd64, arm64, armhf, i386, powerpc, s390x
 racket | 6.11+dfsg1-1       | bionic/universe | source, amd64, arm64, armhf, i386, ppc64el
 racket | 7.2+dfsg1-2ubuntu3 | focal/universe  | source, amd64, arm64, armhf, ppc64el, riscv64, s390x
 racket | 8.2+dfsg1-2ubuntu1 | jammy/universe  | source, amd64, arm64, armhf, ppc64el, riscv64, s390x
 racket | 8.7+dfsg1-1        | mantic/universe | source, amd64, arm64, armhf, ppc64el, riscv64, s390x
 racket | 8.10+dfsg1-2       | noble/universe  | source, amd64, arm64, armhf, ppc64el, riscv64, s390x
```

This means in Ubuntu terms, SWIG only supports the packaged racket version in focal (which is 20.04) and older.  Currently github CI supports 20.04 and 22.04; they dropped 18.04 support 2022-12-01 which suggests 20.04 would probably be dropped around December 2024.

For Debian it's buster (current oldstable) and older.

Given the underlying Scheme implementation has changed, I'd guess the updates required will be significant.  If someone wants to address this it might work better to fork the "mzscheme" target language as a new "racket" target language which only supports racket > 8, leaving "mzscheme" to support < 8 without worrying about breaking that existing support (and then we can drop "mzscheme" once we can no longer regularly test it so the code duplication is a short term thing).  This approach would also fix the SWIG target language name to match the upstream name.

**Comment by erezgeva** [comment]
On Debian
```
$ rmadison racket
racket     | 7.2+dfsg1-2         | oldoldstable             | source, amd64, arm64, armel, armhf, i386, mips, mips64el, mipsel, ppc64el
racket     | 7.9+dfsg1-1~bpo10+1 | buster-backports         | source, amd64, arm64, armel, armhf, i386, mips, mips64el, mipsel, ppc64el
racket     | 7.9+dfsg1-2         | oldstable                | source, amd64, arm64, armel, armhf, i386, mips64el, mipsel, ppc64el, s390x
racket     | 8.7+dfsg1-1~bpo11+1 | bullseye-backports       | source, amd64, arm64, armel, armhf, i386, mips64el, mipsel, ppc64el, s390x
racket     | 8.7+dfsg1-1~bpo11+1 | bullseye-backports-debug | source
racket     | 8.7+dfsg1-1         | stable                   | source, amd64, arm64, armel, armhf, i386, mips64el, mipsel, ppc64el, s390x
racket     | 8.10+dfsg1-2        | testing                  | source, amd64, arm64, armel, armhf, i386, mips64el, ppc64el, s390x
racket     | 8.10+dfsg1-2        | unstable                 | source, amd64, arm64, armel, armhf, i386, mips64el, ppc64el, s390x
racket     | 8.10+dfsg1-2        | unstable-debug           | source
racket     | 8.10+dfsg1-2+b1     | unstable                 | riscv64
racket     | 8.12+dfsg1-1        | experimental             | source, amd64, i386
racket     | 8.12+dfsg1-1        | experimental-debug       | source
racket     | 8.12+dfsg1-3        | buildd-experimental      | source, arm64, ppc64el
racket     | 8.12+dfsg1-3        | experimental             | source, arm64, ppc64el
racket     | 8.12+dfsg1-3        | experimental-debug       | source
```

**Comment by wsfulton** [comment]
Unless someone steps up to update it, perhaps we should remove it altogether in 4.4 and document the intention to do this in the 4.3.0 release notes.

**Comment by ojwb** [comment]
That sounds like a plan.  We're likely to lose the ability to easily test racket in CI in about 3 months time which is likely to be before 4.4.0, and explicit notice of the likely timeline might just connect with someone interested in and able to implement Racket 8 support.

**Comment by erezgeva** [comment]
No objection from my side.
I did notice in `swigmain .cxx` that  "MzScheme/Racket" is marked as Experimental.

**Comment by wsfulton** [comment]
Deprecation notice added in 322bf68.

**Comment by erezgeva** [comment]
> Deprecation notice added in [322bf68](https://github.com/swig/swig/commit/322bf68f396e2d329f36cb36c5f633bc0384035a).

Did we open a ticket in the Racket project?

**Comment by wsfulton** [comment]
> Did we open a ticket in the Racket project?

To report what? 

#920 mentions a new API is now used in Racket 8 and no-one has come forward to basically write a new language module for the new API. That's my understanding, or have I missed something?

**Comment by erezgeva** [comment]
> PI is now used in Racket 8 and no-one has come forward to basically write a new language module for the new API. That's my understanding, or have I missed something?

I mean a ticket (or issue), calling for action.
Asking engineers from the Rocket project to help SWIG support Rocket 8.

**Comment by wsfulton** [comment]
Thanks @erezgeva.

Please see #3038 regarding formal deprecation and removal of target languages.

**Comment by erezgeva** [comment]
https://github.com/gus-massa seems doing some work before:
https://github.com/racket/racket/issues/5077#issuecomment-2390418481

Does anyone care to have a look?

**Comment by erezgeva** [comment]
Hallo [andrewcrook](https://github.com/andrewcrook), [Christopher Bowron](https://github.com/cwbowron), [Demi Marie Obenour](https://github.com/DemiMarie)

You have open an issue on MzScheme in the past.
So we think you might be interesting in supporting Racket/MzScheme.

Currently the maintainers of SWIG are considering the removal of Racket/MzScheme.
We found a work of [Ryan Culpepper](https://github.com/rmculpepper) that support [Racket FFI](https://docs.racket-lang.org/foreign/index.html).
In https://github.com/rmculpepper/swig and document here https://www.cs.umb.edu/~ryanc/tmp/swig/Racket.html.
The project need some one motivated to maintain SWIG's support.
If any of you are interesting, please come forward.
If you know anyone who you think would be interested, please ask him to post here.

**Comment by LiberalArtist** [comment]
I agree that the approach @rmcullpepper recommended in https://github.com/racket/racket/issues/5077#issuecomment-2390418481 would be The Right Thing from a Racket perspective.

I want to clarify that:

> Racket 8 changed the underlying scheme implementation so SWIG doesn't currently work with this version or newer

isn't *strictly* true. In 8.0, the default [implementation](https://docs.racket-lang.org/reference/implementations.html) changed to Racket CS, but the previous default implementation Racket BC (including both the 3M and CGC variants) remains available with [the same C API it's always had](https://docs.racket-lang.org/inside/index.html).

As of [8.13+dfsg1-2](https://salsa.debian.org/bremner/racket/-/commit/a42e83fe8624a2363f12589619b6422aa23e18f7), Debian is still using Racket BC on armel (32-bit soft-float).

**Comment by ojwb** [comment]
Thanks for clarifying.

However we really need to be able to run the testsuite in CI to have any confidence that support for a target language still works, and being used in an armel package doesn't really help there.  Currently the racket packages in the ubuntu-20.04 image work with SWIG, but ubuntu-22.04 and newer racket packages don't so once the ubuntu-20.04 image goes away we lose CI.  That's likely to happen in about 2 months if github follows a similar timeline to ubuntu-18.04's retirement.

SWIG's mzscheme support is also in real need of some dedicated attention - AIUI it doesn't currently even support the 3M variant of BC which was added in 2007.  This code hasn't had a specific maintainer for a really long time (a quick scan of SWIG's git logs seems to show only more global changes touching mzscheme-specific code for 18 years).

Since Racket CS seems to be the future, putting more effort into the existing code doesn't seem useful - better to let it die and instead invest that effort into a new backend which supports Racket the modern way.  Something using FFI would probably make sense, but would need to support wrapping C++ to be accepted as an "experimental" target language (see the manual for the full list of requirements).  Note that the FFI itself doesn't need to support C++ - SWIG could wrap the C++ API as "flat" C functions taking a "this" pointer as the first argument, then call those from racket code using FFI.  If Racket support classes, then those can be created calling the flat functions.  This is how like how wrapping works for some other target languages.

**Comment by LiberalArtist** [comment]
> Since Racket CS seems to be the future, putting more effort into the existing code doesn't seem useful

Agreed. Likewise, if no one steps forward, removing SWIG's mzscheme support seems sensible.

I'm not sure if anyone is actively using the existing SWIG mzscheme support. If such a person exists, they might want to know that they can use Racket 8.0+ however they currently are, they would just need to get or build a non-default `racket`.

If getting CI to work were just a matter of needing to install BC, https://github.com/Bogdanp/setup-racket/ could help with it. (But you could also reasonably decide that it's still not worth it!) If CGC is really needed (I'm not 100% clear on that), that would be more challenging, but is probably no different than the CI situation now.

**Comment by ojwb** [comment]
If I understood #920 correctly, SWIG-generated bindings may be currently broken with the BC garbage collector (it seems we need to register all C variables that point to Scheme garbage-collected value, but I didn't find any evidence of a change in SWIG to implement that).

The testsuite currently passes in CI but that might just mean the testcases are too small for GC to be needed.

**Comment by LiberalArtist** [comment]
I am not confident that I understand everything SWIG currently does correctly, either, but [`raco ctool --xform`](https://docs.racket-lang.org/inside/im_memoryalloc.html#%28part._im~3a3m~3amzc%29) can transform C source code that works with the CGC garbage collector to automatically add the annotations needed for 3M, the precise garbage collector that is the default for Racket BC. While the documentation focuses on the limitations of the tool, it is powerful enough to transform the C source for the Racket BC implementation itself (for bootstrapping), so I imagine it would also be viable for SWIG-generated C code.

**Comment by ojwb** [comment]
Github are retiring support for ubuntu-20.04 in GHA - actual EOL is 2025-04-01 (so a month away as I write) but they're going to be doing "brownouts" on and off during March (see https://github.com/actions/runner-images/issues/11101) so I've started work on upgrading our CI to not use ubuntu-20.04.  See #3132.

Our current CI testing for Racket requires a version available as a package in 20.04 but not 22.04.  No CI testing means we can't know that Racket support still works in future releases and that means Racket support is likely to get removed entirely unless somebody is actually actively working on resolving this.  If that's you then you need to speak up very soon!

**Comment by vadz** [comment]
FWIW it's quite simple to keep using Ubuntu 20.04 by running the job on `ubuntu-latest` image but specifying `ubuntu:20.04` as the container to use for it. It's not a long-term solution, of course, but could be used to postpone the need to do something about this immediately.

**Comment by ojwb** [comment]
Note that it's already been two months more than the approximate date I predicted for GHA retiring ubuntu-20.04 in previous comments here.

We've already postponed doing anything about this several times and at this point I really don't see any point delaying what now seems inevitable, unless somebody is actually already actively working on this and just needs a bit more time (but even then I'm not sure it helps them to keep support for an ancient Racket version in the tree as this seem like it's going to require major changes; probably the most promising link above was a fork of SWIG switched to use Racket's FFI but that would need extending to support C++ because Racket's FFI only supports C - probably SWIG could generate C flat functions to call the C++ methods and wrap those with Racket's FFI).

End of standard support for Ubuntu 20.04 is 2025-05-29 anyway (there's extended support until 2030 but you have to pay to get access to that).

We also haven't had a maintainer for SWIG's mzscheme/racket support for a long time, and someone above (who I assume is from the Racket community) doubted it had any active users either:

> I'm not sure if anyone is actively using the existing SWIG mzscheme support.

We don't have enough developer time to even properly keep up with incoming bug reports which are (presumably) from active users, so let's not divert any to keeping the Racket support on life support for just a few more months.
